### PR TITLE
sis

### PR DESCRIPTION
**Pull request by abdelrahman0w**

### CODE REVIEW

File: ej_cli/sis/regex.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+matches = {
```
Comment (aboueleyes) [inline ej_cli/sis/regex.py:1]: wouldn't it be better to compile it so as not to do it every time?
```python
re.compile(r'a regex')
````

File: ej_cli/sis/sis.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    def __init__(self, id: str, pwd: str, save_credentials: bool = True, new_pwd: bool = False) -> None:
+        self.id = id
+        self.pwd = pwd
+        self.save_credentials = save_credentials
+        self.__data = {}
+        self.__session = None
+        self.__authenticated = False
+
+        if new_pwd and os.path.exists(
+            os.path.join(PARENT_DIR, f"saved/{self.id}-credentials.json")
```
Comment (aboueleyes) [inline ej_cli/sis/sis.py:26]: use `os.path.join` also don't use `/` or use `os.sep` to be windows compatible

File: ej_cli/sis/sis.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        ):
+            os.remove(os.path.join(
+                PARENT_DIR, f"saved/{self.id}-credentials.json"))
```
Comment (aboueleyes) [inline ej_cli/sis/sis.py:29]: same here
and since this is duplicated make it a property

File: ej_cli/sis/sis.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+    def open_session(self) -> requests.Session:
+        if not self.__session:
+            self.__session = requests.Session()
+
+        return self.__session
+
+    def close_session(self, req: requests.Session) -> None:
+        req.close()
+        self.session = None
```
Comment (aboueleyes) [inline ej_cli/sis/sis.py:49]: context manager would be better?

File: ej_cli/sis/sis.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+            "__EVENTARGUMENT": "",
+            "__VIEWSTATE": VIEWSTATE,
+            "__VIEWSTATEGENERATOR": VIEWSTATEGENERATOR,
+            "__EVENTVALIDATION": EVENTVALIDATION,
+            "txtUsername": self.id,
+            "txtPassword": self.pwd,
+            "btnEnter": "Log In"
+        }
+        self.req(
+            "https://sis.ejust.edu.eg/Default.aspx", method="POST", data=login_data
```
Comment (aboueleyes) [inline ej_cli/sis/sis.py:74]: extract URL to constant b2a

File: ej_cli/sis/sis.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+                ["Student ID", self.__data["student_id"]],
+                ["CGPA", self.__data["cgpa"]],
+            ],
+            tablefmt="rst",
+        )
+
+    @property
+    def credit_hours(self) -> str:
+        if not self.__data:
+            self.__get_info()
```
Comment (aboueleyes) [inline ej_cli/sis/sis.py:194]: decorator b2a would be cooler?

File: ej_cli/sis/sis.py (hunk 4 above)
Comment (Abdelrahman0W) [reply ej_cli/sis/sis.py:49]: Unfortunately not, the requests being sent to the website take a while. Also, all pages can't be accessed without auth. I tried it before doing this method and it didn't work. SIS sucks :disappointed:

File: ej_cli/sis/regex.py (hunk 1 above)
Comment (Abdelrahman0W) [reply ej_cli/sis/regex.py:1]: I will try it :100:

File: ej_cli/sis/sis.py (hunk 5 above)
Comment (Abdelrahman0W) [reply ej_cli/sis/sis.py:74]: :+1: 
Also, I will add new issues if you want to contribute :heart:

File: ej_cli/sis/sis.py (hunk 4 above)
Comment (Abdelrahman0W) [reply ej_cli/sis/sis.py:49]: I mean it won't be the best experience to authenticate everytime.

Comment (Abdelrahman0W) [reply ej_cli/sis/sis.py:49]: But I will try to figure out a way to do it

**Pull request merged**